doc/hanging-socket.md

```diff
@@ -0,0 +1,99 @@
+                            seastar::stop_iteration::yes);
+                }
+            });
+        }).then([&out] {
+            return out.close();
+        });
+    });
+}
+```
+In this example, the next step after read, will never happen, and we'll never reach the "end" state.
+
+A simple solution for this case might be to add a timeout on the read operation, in the case the client doesn't do any progress, for instance 
+```cpp
+seastar::future<> handle_connection(seastar::connected_socket s) {
+    auto out = s.output();
+    auto in = s.input();
+
+    return do_with(std::move(s), std::move(out), std::move(in),
+            [] (auto& s, auto& out, auto& in) {
+        return seastar::repeat([&out, &in] {
+            return seastar::with_timeout(seastar::lowres_clock::now() + std::chrono::seconds(1), in.read()).then([&out] (auto buf) {
+                if (buf) {
+                    return out.write(std::move(buf)).then([&out] {
+                        return out.flush();
+                    }).then([] {
+                        return seastar::stop_iteration::no;
+                    });
+                } else {
+                    return seastar::make_ready_future<seastar::stop_iteration>(
+                            seastar::stop_iteration::yes);
+                }
+            });
+        }).then([&out] {
+            return out.close();
+        });
+    });
+}
+```
+In this example, a timeout exception will be propogonated, but there is a problem with that, ```seastar::with_timeout```, doesn't cancel ```in.read()``` future, which
+means, the future might run, and can cause a Segmentation Fault, due to the fact that the ```sesastar::input_stream<char>```, was already destructed.
+
+In order to resolve such a case, we can add a unique handling for the case of timeout
+```cpp
+seastar::future<> handle_connection(seastar::connected_socket s) {
+    auto out = s.output();
+    auto in = s.input();
+
+    return do_with(std::move(s), std::move(out), std::move(in),
+            [] (auto& s, auto& out, auto& in) {
+        return seastar::repeat([&out, &in] {
+            return seastar::with_timeout(seastar::lowres_clock::now() + std::chrono::seconds(1), in.read()).then([&out] (auto buf) {
+                if (buf) {
+                    return out.write(std::move(buf)).then([&out] {
+                        return out.flush();
+                    }).then([] {
+                        return seastar::stop_iteration::no;
+                    });
+                } else {
+                    return seastar::make_ready_future<seastar::stop_iteration>(
+                            seastar::stop_iteration::yes);
+                }
+            }).handle_exception_type([&s] (const seastar::timed_out_error &err) {
+                // Will resolve the currently waiting futures, with an exception
+                s.shutdown_input();
+                // Allows to wait untill the inputs are actually closed
```

Comment:
Well this is also wrong, since the original continuation escapes. Maybe it works for the example, but for the general case, it's problematic.

The correct approach is to have a separate monitoring fiber issue the shutdown when it wants. Then the original write() just terminates with an exception. It's also  more efficient as we don't wrap everything with with_timeout wrappers.